File: javascript/functional.js
#!/usr/bin/env node

// 'functional.js'.
// Chris Shiels.


'use strict';


const all = function(f, l) {
  if (l.length === 0)
    return false;
  else if (l.length === 1)
    return f(l[0]);
  else
    return f(l[0]) && all(f, l.slice(1));
}


const any = function(f, l) {
  if (l.length === 0)
    return false;
  else if (l.length === 1)
    return f(l[0]);
  else
    return f(l[0]) || any(f, l.slice(1));
}


const flatten = function(l) {
  if (l.length === 0)
    return [];
  else if (l[0] instanceof Array)
    return flatten(l[0]).concat(flatten(l.slice(1)));
  else
    return [ l[0] ].concat(flatten(l.slice(1)));
}


const range = function(m, n = null, s = 1) {
  const internal = function(start, stop, step) {
    if (step === 0 ||
        (step > 0 && start >= stop) ||
        (step < 0 && start <= stop))
      return [];
    else
      return [ start ].concat(internal(start + step, stop, step));
  }
  if (n === null)
    return internal(0, m, s);
  else
    return internal(m, n, s);
}


const reduce = function(f, l, v = null) {
  const internal = function(f, l, v) {
    if (l.length === 0)
      return v;
    else
      return internal(f, l.slice(1), f(v, l[0]));
  }
  if (v === null)
    return internal(f, l.slice(1), l[0]);
  else
    return internal(f, l, v);
}


const scan = function(f, l, v) {
  const internal = function(f, l, a) {
    if (l.length === 0)
      return a;
    else
      return internal(f, l.slice(1), a.concat(f(a.slice(-1)[0], l[0])));
  }
  return internal(f, l, [ v ]);
}


const map = function(f, l) {
  const accumulate = function(a, e) {
    return a.concat([ f(e) ]);
  }
  return reduce(accumulate, l, []);
}


const filter = function(f, l) {
  const accumulate = function(a, e) {
    if (f(e))
      return a.concat([ e ]);
    else
      return a;
  }
  return reduce(accumulate, l, []);
}


const partition = function(f, l) {
  const accumulate = function(a, e) {
    let left, right;
    [ left, right ] = a;
    if (f(e))
      return [ left.concat([ e ]), right ];
    else
      return [ left, right.concat([ e ]) ];
  }
  return reduce(accumulate, l, [ [], [] ]);
}


const split = function(f, l) {
  const accumulate = function(a, e) {
    if (f(e))
      return a.concat([ [] ]);
    else {
      return a.slice(0, -1).concat([ a.slice(-1)[0].concat(e) ]);
    }
  }

  const internal = function(e) {
    return e.length !== 0;
  }

  return filter(internal, reduce(accumulate, l, [ [] ]));
}


const reverse = function(l) {
  const accumulate = function(a, e) {
    return [ e ].concat(a);
  }
  return reduce(accumulate, l, []);
}


const sort = function(f, l) {
  if (l.length === 0)
    return [];
  return sort(f, filter((e) => { return f(e, l[0]); }, l.slice(1)))
           .concat([ l[0] ])
           .concat(sort(f, filter((e) => { return !f(e, l[0]); }, l.slice(1))));
}


const unique = function(l) {
  const accumulate = function(a, e) {
    if (a.includes(e))
      return a;
    else
      return a.concat([ e ]);
  }
  return reduce(accumulate, l, []);
}


const zip = function(l1, l2) {
  if (l1.length === 0 || l2.length === 0)
    return [];
  else
    return [ (l1[0], l2[0]) ].concat(zip(l1.slice(1), l2.slice(1)));
}


const zipwith = function(f, l1, l2) {
  if (l1.length === 0 || l2.length === 0)
    return [];
  else
    return [ f(l1[0], l2[0]) ].concat(zipwith(f, l1.slice(1), l2.slice(1)));
}


const group = function(l) {
  const accumulate = function(a, e) {
    if (a.length && a.slice(-1)[0].includes(e))
      return a.slice(0, -1).concat([ a.slice(-1)[0].concat([ e ]) ]);
    else
      return a.concat([ [ e ] ]);
  }
  return reduce(accumulate, l, []);
}


const permutations = function(l) {
  if (l.length === 0)
    return [];
  else if (l.length === 1)
    return [ l ];
  else
    return reduce((a, e1) =>
	            { return a.concat(map((e2) =>
		                        { return [ e1 ].concat(e2) },
	                                    permutations(filter((e3) =>
						           { return e3 != e1 },
						         l)))) },
		   l,
	           []);
}


const flip = function(f) {
  return function(...args) {
    return f(...reverse(args));
  }
}


const compose = function(f1, f2) {
  return function(v) {
    return f2(f1(v));
  }
}


const pipe = function(l) {
  const accumulate = function(a, e) {
    return e(a);
  }
  return function(v) {
    return reduce(accumulate, l, v);
  }
}


const pipe2 = function(l) {
  return reduce(compose, l);
}


const pipemaybe = function(l) {
  const accumulate = function(a, e) {
    if (a !== null)
      return e(a);
    else
      return a;
  }
  return function(v) {
    return reduce(accumulate, l, v);
  }
}


const pipeeither = function(l) {
  const accumulate = function(a, e) {
    if (a[0] === null)
      return e(a[1]);
    else
      return a;
  }
  return function(v) {
    return reduce(accumulate, l, [null, v]);
  }
}


const partial = function(f, ...args1) {
  return function(...args2) {
    return f(...args1.concat(args2));
  }
}


const curry = function(f, arity) {
  arity = arity || f.length;

  return function(v) {
    if (arity === 1)
      return f(v);
    else
      return curry(partial(f, v), arity - 1);
  }
}


const memoize = function(f) {
  let cache = {};

  return function(...args) {
    let key = JSON.stringify(args);
    if (!(key in cache))
      cache[key] = f(...args);
    return cache[key];
  }
}


const factorial = function(n) {
  if (n <= 1)
    return 1;
  else
    return n * factorial(n - 1);
}


const fibonacci = function(n) {
  if (n === 1 || n === 2)
    return 1;
  else
    return fibonacci(n - 1) + fibonacci(n - 2);
}


const memoizedfibonacci = function(n) {
  let fibonacci = function(n) {
    if (n === 1 || n === 2)
      return 1;
    else
      return fibonacci(n - 1) + fibonacci(n - 2);
  }
  fibonacci = memoize(fibonacci);
  return fibonacci(n);
}


const tablefibonacci = function(n) {
  const fibonacci = function(a, e) {
    if (e === 1 || e === 2)
      return a.concat([ 1 ]);
    else
      return a.concat([ a.slice(-1)[0] + a.slice(-2)[0] ]);
  }
  return reduce(fibonacci, range(1, n + 1), []).slice(-1);
}


const primes = function(n) {
  const sieve = function(l) {
    if (l.length === 0)
      return [];
    else
      return [ l[0] ].concat(sieve(filter((e) => { return e % l[0] != 0 },
                                          l.slice(1))));
  }

  return sieve(range(2, n));
}


const ispalindrome = function(s) {
  if (s.length === 0)
    return true;
  else if (s.slice(0, 1) != s.slice(-1))
    return false;
  else
    return ispalindrome(s.slice(1, -1));
}


module.exports = {
  'all':               all,
  'any':               any,
  'flatten':           flatten,
  'range':             range,
  'reduce':            reduce,
  'scan':              scan,
  'map':               map,
  'filter':            filter,
  'partition':         partition,
  'split':             split,
  'reverse':           reverse,
  'sort':              sort,
  'unique':            unique,
  'zip':               zip,
  'zipwith':           zipwith,
  'group':             group,
  'permutations':      permutations,
  'flip':              flip,
  'compose':           compose,
  'pipe':              pipe,
  'pipe2':             pipe2,
  'pipemaybe':         pipemaybe,
  'pipeeither':        pipeeither,
  'partial':           partial,
  'curry':             curry,
  'memoize':           memoize,
  'factorial':         factorial,
  'fibonacci':         fibonacci,
  'memoizedfibonacci': memoizedfibonacci,
  'tablefibonacci':    tablefibonacci,
  'primes':            primes,
  'ispalindrome':      ispalindrome
};

File: python/functional.py
#!/usr/bin/env python

# 'functional.py'.
# Chris Shiels.


def all(f, l):
  if len(l) == 0:
    return False
  elif len(l) == 1:
    return f(l[0])
  else:
    return f(l[0]) and all(f, l[1:])


def any(f, l):
  if len(l) == 0:
    return False
  elif len(l) == 1:
    return f(l[0])
  else:
    return f(l[0]) or any(f, l[1:])


def flatten(l):
  if l == []:
    return []
  elif type(l[0]) is list:
    return flatten(l[0]) + flatten(l[1:])
  else:
    return [ l[0] ] + flatten(l[1:])


def range(m, n = None, s = 1):
  def internal(start, stop, step):
    if step == 0 or \
       (step > 0 and start >= stop) or \
       (step < 0 and start <= stop):
      return []
    else:
      return [ start ] + internal(start + step, stop, step)
  if n is None:
    return internal(0, m, s)
  else:
    return internal(m, n, s)


def reduce(f, l, v = None):
  def internal(f, l, v):
    if l == []:
      return v
    else:
      return internal(f, l[1:], f(v, l[0]))
  if v is None:
    return internal(f, l[1:], l[0])
  else:
    return internal(f, l, v)


def scan(f, l, v):
  def internal(f, l, a):
    if l == []:
      return a
    else:
      return internal(f, l[1:], a + [f(a[-1], l[0])])
  return internal(f, l, [ v ])


def map(f, l):
  def accumulate(a, e):
    return a + [ f(e) ]
  return reduce(accumulate, l, [])


def filter(f, l):
  def accumulate(a, e):
    if f(e):
      return a + [ e ]
    else:
      return a
  return reduce(accumulate, l, [])


def partition(f, l):
  def accumulate(a, e):
    left, right = a
    if f(e):
      return [ left + [ e ], right ]
    else:
      return [ left, right + [ e ] ]
  return reduce(accumulate, l, [ [], [] ])


def split(f, l):
  def accumulate(a, e):
    if f(e):
      return a + [ [] ]
    else:
      return a[0:-1] + [ a[-1] + [ e ] ]

  def internal(e):
    return e != []
  return filter(internal, reduce(accumulate, l, [ [] ]))


def reverse(l):
  def accumulate(a, e):
    return [ e ] + a
  return reduce(accumulate, l, [])


def sort(f, l):
  if l == []:
    return []
  else:
    return sort(f, filter(lambda e: f(e, l[0]), l[1:])) + \
           [ l[0] ] + \
           sort(f, filter(lambda e: not f(e, l[0]), l[1:]))


def unique(l):
  def accumulate(a, e):
    if e in a:
      return a
    else:
      return a + [ e ]
  return reduce(accumulate, l, [])


def zip(l1, l2):
  if l1 == [] or l2 == []:
    return []
  else:
    return [ (l1[0], l2[0]) ] + zip(l1[1:], l2[1:])


def zipwith(f, l1, l2):
  if l1 == [] or l2 == []:
    return []
  else:
    return [ f(l1[0], l2[0]) ] + zipwith(f, l1[1:], l2[1:])


def permutations(l):
  if len(l) == 0:
    return []
  elif len(l) == 1:
    return [ l ]
  else:
    return reduce(lambda a, e1: a + map(lambda e2: [ e1 ] + e2,
                                        permutations(filter(lambda e3: e3 != e1,
                                                            l))),
                  l,
                  [])


def compose(f1, f2):
  def internal(v):
    return f2(f1(v))
  return internal


def pipe(l):
  def accumulate(a, e):
    return e(a)

  def internal(v):
    return reduce(accumulate, l, v)
  return internal


def pipe2(l):
  return reduce(compose, l)


def pipemaybe(l):
  def accumulate(a, e):
    if a != None:
      return e(a)
    else:
      return None

  def internal(v):
    return reduce(accumulate, l, v)
  return internal


def partial(f, *args1):
  def internal(*args2):
    return f(*(args1 + args2))
  return internal


def curry(f, arity = 0):
  arity = arity or f.__code__.co_argcount

  def internal(v):
    if arity == 1:
      return f(v)
    else:
      return curry(partial(f, v), arity - 1)
  return internal


def memoize(f):
  cache = {}

  def internal(*args):
    key = str(args)
    if not key in cache:
      cache[key] = f(*args)
    return cache[key]
  return internal


def factorial(n):
  if n <= 1:
    return 1
  else:
    return n * factorial(n - 1)


def fibonacci(n):
  if n <= 1:
    return 1
  else:
    return fibonacci(n - 1) + fibonacci(n - 2)


def memoizedfibonacci(n):
  def fibonacci(n):
    if n <= 1:
      return 1
    else:
      return fibonacci(n - 1) + fibonacci(n - 2)
  fibonacci = memoize(fibonacci)
  return fibonacci(n)


def primes(n):
  def sieve(l):
    if l == []:
      return []
    else:
      return [ l[0] ] + sieve(filter(lambda e: e % l[0] != 0, l[1:]))
  return sieve(range(2, n))


def ispalindrome(s):
  if len(s) == 0:
    return True
  elif s[0] != s[-1]:
    return False
  else:
    return ispalindrome(s[1:-1])
